### PR TITLE
rgw: [CORTX-32034] Fixed same Sid issue in put-user-policy operation

### PR DESCRIPTION
**Pull request by sanalpk**
**Problem** : Same SID's allowed in put-user-policy operation

**Root Cause**: There is no check for detecting same SIDs in a policy document

**Solution** : Introduced a new set sid_set and adding each sid to it.
                       If sid is already present, we are throwing error "MalformedPolicyDocument"

Signed-off-by: Sanal Kaarthikeyan 

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by sachinpunadikar — APPROVED** [review]
LGTM

File: src/rgw/rgw_iam_policy.cc [hunk 1]
```diff
@@ -164,6 +164,7 @@ static const actpair actpairs[] =
 
 struct PolicyParser;
 
+static std::set<std::string> sid_set;
```
Comment (sumedhak27) [inline src/rgw/rgw_iam_policy.cc:167]: Let's use `unordered_set` since we don't want sid's in sorted order.

File: src/rgw/rgw_iam_policy.cc [hunk 2]
```diff
@@ -512,6 +513,12 @@ bool ParseState::do_string(CephContext* cct, const char* s, size_t l) {
     // Statement
 
   } else if (w->id == TokenID::Sid) {
+    std::pair<std::set<std::string>::iterator,bool> ret;
+    std::string sid(s, l);
+    ret = sid_set.insert(sid);
+    if (ret.second == false) {
+      return false;
+    }
```
Comment (sumedhak27) [inline src/rgw/rgw_iam_policy.cc:521]: Current code looks little intimidating to me. 😅, else it is fine.
```c++
    std::string sid(s, l);
    if (sid_set.insert(sid).second == false) {
      return false;
    }
```

**Review by jjxsg — APPROVED** [review]
> Solution : Added logic to detect same SIDs in a policy document

Instead of saying added some logic, please always add explanation of actual logic implemented.

**Review by bkirunge7 — APPROVED** [review]

File: src/rgw/rgw_iam_policy.cc [hunk 3]
```diff
@@ -512,6 +513,11 @@ bool ParseState::do_string(CephContext* cct, const char* s, size_t l) {
     // Statement
 
   } else if (w->id == TokenID::Sid) {
+      std::string sid(s, l);
+      if (sid_set.find(sid) != sid_set.end()) {
+        return false;
+      }
```
Comment (sumedhak27) [inline src/rgw/rgw_iam_policy.cc:519]: indentation is off by 2 spaces @sanalpk..

File: src/rgw/rgw_iam_policy.cc [hunk 4]
```diff
@@ -512,6 +513,11 @@ bool ParseState::do_string(CephContext* cct, const char* s, size_t l) {
...
+    sid_set.insert(sid);
```
Comment (sumedhak27) [inline src/rgw/rgw_iam_policy.cc:520]:
```suggestion
    std::string sid(s, l);
    if (sid_set.find(sid) != sid_set.end()) {
      return false;
    }
    sid_set.insert(sid);
```

Comment (sanalpk) [reply src/rgw/rgw_iam_policy.cc]: Done

**Review by sumedhak27 — APPROVED** [review]

**Pull request merged**